### PR TITLE
Steel effect with auto-framing and associated typechecker changes

### PR DESCRIPTION
**Pull request by aseemr**
I now have small examples working for the new steel effect that includes auto-framing. The examples don't need an explicit return either.

The effect is in `ulib/experimental/Steel.FramingEffect.fst` and tests in `examples/steel/`Steel.Effects2.Tests.fst`.

Main typechecker changes include:

* `polymonadic_subcomp`, specifically to allow `SteelF <: Steel`
* Returning pure computations into `M` when `(M, PURE)` cannot be composed, this is done as part of `maybe_return_e2_and_bind`.

I already pushed some related changes into master. These changes include using `M.return` rather than `PURE.return` and then `lift_PURE_M` when lifting `e2` in `maybe_return_e2_and_bind`. This combined with the second bullet above gives us a generic implementation of `bind_M_PURE`.

### CODE REVIEW

File: src/syntax/FStar.Syntax.Util.fs [hunk 1]
```diff
@@ -250,6 +250,25 @@ let comp_to_comp_typ (c:comp) : comp_typ =
...
+ * For wp effects, given a (unit -> M a wp), return wp
+ *
+ * The pattern matching is very syntactic inside this function
+ *)
+let effect_indices_from_repr (repr:term) (is_layered:bool) (r:Range.range) (err:string)
+: list<term> =
+  let err () = Errors.raise_error (Errors.Fatal_UnexpectedEffect, err) r in
+  let repr = compress repr in
+  if is_layered
+  then match repr.n with
```
Comment (nikswamy) [inline src/syntax/FStar.Syntax.Util.fs:265]: Does the representation have to be an app node for a layered effect? I could use a arrow literal, right?
Also, maybe it is safer to use U.head_and_args in the this case, rather than directly matching on a Tm_app

Likewise, in the non-layered case, does the repr have to be an arrow? And to use arrow_formals_comp to destruct it, if it is an arrow.

File: src/typechecker/FStar.TypeChecker.DeferredImplicits.fs [hunk 2]
```diff
@@ -136,7 +136,7 @@ let solve_deferred_to_tactic_goals env g =
         in
         let goal_ty = U.mk_eq2 (env.universe_of env_lax t_eq) t_eq tp.lhs tp.rhs in
         let goal, ctx_uvar, _ =
-            Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Strict None
+            Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Allow_untyped None
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.DeferredImplicits.fs:139]: Is this unintentional? I changed it to Strict on your suggestion and this commit may be reverting that by mistake.

File: src/typechecker/FStar.TypeChecker.TcEffect.fs [hunk 3]
```diff
@@ -460,6 +460,20 @@ let tc_layered_eff_decl env0 (ed : S.eff_decl) (quals : list<qualifier>) =
...
+     *   we need to adjust the qualifiers
+     *
+     * Implicits remain implicits, but meta_attr or meta_arg just become implicits
+     *
+     * Don't think the boolean true below matters
+     *)
+    let binder_aq_to_arg_aq aq =
+      match aq with
+      | Some (Implicit _) -> aq
+      | Some (Meta _) -> Some (Implicit true)
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.TcEffect.fs:474]: probably irrelevant, but may be safer to use Implicit false.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 4]
```diff
@@ -1085,14 +1085,20 @@ and tc_value env (e:term) : term
...
-                   TcUtil.new_implicit_var "type of user-provided implicit term" r env k
+        | None ->
+          let k, u = U.type_u () in
+          TcUtil.new_implicit_var "type of user-provided implicit term" r env k
         | Some t -> t, [], Env.trivial_guard in
-    let e, _, g1 = TcUtil.new_implicit_var ("user-provided implicit term at " ^ (Range.string_of_range r)) r env t in
+    
+    let e, _, g1 = TcUtil.new_implicit_var
+          ("user-provided implicit term at " ^ (Range.string_of_range r))
+          r env t in
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:1101]: Looks to be only formatting and comment change here. Am I missing something

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 5]
```diff
@@ -3847,6 +3876,55 @@ let teq_nosmt (env:env) (t1:typ) (t2:typ) : option<guard_t> =
...
+ * -- leaving their typechecking to resolve_implicits'
+ *
+ * E.g. for a ?u:squash phi, this will only set ?u=unit in the UF graph,
+ *   and, as usual, resolve_implicits' will check that G |= phi
+ *
+ * It returns a boolean (true if at least one implicit was solved)
+ *   and the set of new implicits, right now this set is same as imps,
+ *   for inductives, this may later include implicits for pattern variables
+ *)
+
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:3898]: This part is from PR #2041, right? Can we merge that separately?

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 6]
```diff
@@ -3200,113 +3198,140 @@ and solve_c (env:Env.env) (problem:problem<comp>) (wl:worklist) : solution =
...
+            Env.pure_precondition_for_trivial_post env u stronger_ct.result_typ wp Range.dummyRange in
+
+          let sub_probs = ret_sub_prob::(is_sub_probs@f_sub_probs@g_sub_probs@(g_lift.deferred |> List.map snd)) in
+          let guard =
+            let guard = U.mk_conj_l (List.map p_guard sub_probs) in
+            match g_lift.guard_f with
+            | Trivial -> guard
+            | NonTrivial f -> U.mk_conj guard f in
+          let wl = solve_prob orig (Some <| U.mk_conj guard fml) [] wl in
+          solve env (attempt sub_probs wl) in
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:3334]: This is a big diff. Can you summarize the main intent of the change here?

File: src/syntax/FStar.Syntax.Util.fs (hunk 1 above)
Comment (aseemr) [reply src/syntax/FStar.Syntax.Util.fs:265]: This function is expected to be called with the shape that it matches on (there is also a comment on the function). The idea is to be able to parse the effect indices given a `repr` term that is exactly `repr a is` for layered effects and `unit -> M a wp` for `wp` effects.

The way we store the layered effect combinator types in the env, including lifts, it is already the case that computation arguments (`f`, `g`, etc.) to those combinators have exactly these shapes.

File: src/typechecker/FStar.TypeChecker.DeferredImplicits.fs (hunk 2 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.DeferredImplicits.fs:139]: Reverted it, thanks for pointing out!

File: src/typechecker/FStar.TypeChecker.TcEffect.fs (hunk 3 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.TcEffect.fs:474]: Done.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs (hunk 4 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.TcTerm.fs:1101]: Yeah it is just formatting, the line was too long.

File: src/typechecker/FStar.TypeChecker.Rel.fs (hunk 5 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.Rel.fs:3898]: It would be tricky, I have made some changes to this code ... e.g. not having this code kick in for tactics. If you have some comments on that PR, I will be happy to accommodate them here, but it would be easier for me to merge it as part of this PR rather than porting changes to that branch and then do it

File: src/typechecker/FStar.TypeChecker.Rel.fs (hunk 6 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.Rel.fs:3334]: Not much actually. This code is tweaked a little to support `polymonadic_subcomp`. Earlier to solve a `c1 <: c2`, we would first lift `c1` to `c2` or vice versa, and then call either `solve_sub` or `solve_layered_sub` depending on whether the resultant effect is layered.

But now, such a lift may not exist, and instead there may be a polymoandic subcomp. So, for `c1 <: c2`, this code says if either is a layered effect go to `solve_layered_sub` which will decide whether to lift or use a polymonadic subcomp.

File: src/syntax/FStar.Syntax.Util.fs (hunk 1 above)
Comment (aseemr) [reply src/syntax/FStar.Syntax.Util.fs:265]: I added further comments to this function.

**Comment by aseemr** [comment]
Many thanks @nikswamy for the review! Please see my comments. If we are happy with it, should I merge it in?

**Comment by nikswamy** [comment]
hi @aseemr  thanks for all your responses. I'm fine with the merge once you're happy with CI.
